pkg/helm/update_release.go

```diff
@@ -0,0 +1,120 @@
+package helm
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+
+	"github.com/pkg/errors"
+	"helm.sh/helm/v3/pkg/release"
+	corev1 "k8s.io/api/core/v1"
+	kuberneteserrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	HelmReleaseSecretType = "helm.sh/release.v1"
+)
+
+// MigrateExistingHelmReleaseSecrets will move all helm release secrets from the kotsadm namespace to the release namespace
+func MigrateExistingHelmReleaseSecrets(clientset kubernetes.Interface, releaseName string, releaseNamespace string, kotsadmNamespace string) error {
+	selectorLabels := labels.Set{
+		"owner": "helm",
+		"name":  releaseName,
+	}.AsSelector()
+	fieldSelectorMap := fields.Set{
+		"type": HelmReleaseSecretType,
+	}.AsSelector()
+	listOpts := metav1.ListOptions{
+		LabelSelector: selectorLabels.String(),
+		FieldSelector: fieldSelectorMap.String(),
+	}
+
+	secretList, err := clientset.CoreV1().Secrets(kotsadmNamespace).List(context.TODO(), listOpts)
+	if err != nil {
+		if kuberneteserrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list release secrets for %s", releaseName)
+	}
+
+	if len(secretList.Items) == 0 {
+		return nil
+	}
+
+	for _, secret := range secretList.Items {
+		if secret.Namespace != releaseNamespace {
+			err := moveHelmReleaseSecret(clientset, secret, releaseNamespace)
+			if err != nil {
+				return errors.Wrapf(err, "failed to move helm release secret %s to %s", secret.Name, releaseNamespace)
+			}
+		}
+	}
+	return nil
+}
+
+// moveHelmReleaseSecret will create a new secret in the releaseNamespace and delete the old one from the kotsadmNamespace
+func moveHelmReleaseSecret(clientset kubernetes.Interface, secret corev1.Secret, releaseNamespace string) error {
+	release, err := HelmReleaseFromSecretData(secret.Data["release"])
+	if err != nil {
+		return errors.Wrapf(err, "failed to get release from secret data")
+	}
+
+	// set release namespace to the new namespace
+	release.Namespace = releaseNamespace
+	releaseStr, err := encodeRelease(release)
+	if err != nil {
+		return errors.Wrapf(err, "failed to encode release")
+	}
+
+	newReleaseSecret := corev1.Secret{
```

Comment:
Hmm, any reason to not just use the existing secret? Who knows what Helm will add to it later. That also seems to be how this article did it https://blog.random.io/moving-helm-release-to-another-kubernetes-namespace/